File: sidecar/app/session/domain/services/session_service_impl.py
"""Session domain service implementation."""

from __future__ import annotations

from sidecar.app.session.domain.abstractions.session_conversation_manager import SessionConversationManager
from sidecar.app.session.domain.abstractions.session_repository import SessionRepository
from sidecar.app.session.domain.abstractions.session_service import SessionService
from sidecar.app.session.domain.dtos.create_session_dto import CreateSessionDto
from sidecar.app.session.domain.dtos.update_session_dto import UpdateSessionDto
from sidecar.app.session.domain.entities.session_entity import SessionEntity
from sidecar.shared.exceptions import NotFoundException


class SessionServiceImpl(SessionService):
    """Default session service backed by a session repository."""

    def __init__(
        self,
        session_repository: SessionRepository,
        session_conversation_manager: SessionConversationManager,
    ) -> None:
        self._session_repository = session_repository
        self._session_conversation_manager = session_conversation_manager

    async def create(self, sessionDto: CreateSessionDto) -> SessionEntity:
        session_entity = SessionEntity(
            name=sessionDto.session_name,
            preset_id=sessionDto.preset_id,
        )
        created_session = await self._session_repository.create(session_entity)
        try:
            await self._session_conversation_manager.create(
                session_id=created_session.id,
                preset_id=created_session.preset_id,
            )
        except Exception:
            await self._session_repository.delete_by_id(created_session.id)
            raise
        return created_session

    async def get_one_by_id(self, session_id: str) -> SessionEntity:
        session_entity = await self._session_repository.get_one_by_id(session_id)
        if session_entity is None:
            raise NotFoundException(f"Session not found: {session_id}")
        return session_entity

    async def get(self) -> list[SessionEntity]:
        return await self._session_repository.get()

    async def update_by_id(self, session_id: str, updateDto: UpdateSessionDto) -> SessionEntity:
        existing_session = await self.get_one_by_id(session_id)
        if updateDto.session_name is not None:
            existing_session.name = updateDto.session_name

        updated_session = await self._session_repository.update_by_id(session_id, existing_session)
        if updated_session is None:
            raise NotFoundException(f"Session not found: {session_id}")
        return updated_session

    async def end_by_id(self, session_id: str) -> SessionEntity:
        existing_session = await self.get_one_by_id(session_id)
        existing_session.end()

        updated_session = await self._session_repository.update_by_id(session_id, existing_session)
        if updated_session is None:
            raise NotFoundException(f"Session not found: {session_id}")
        await self._session_conversation_manager.close(session_id)
        return updated_session

    async def delete_by_id(self, session_id: str) -> None:
        await self._session_repository.delete_by_id(session_id)

File: sidecar/src/app/session/domain/services/session-service-impl.ts
import type { SessionAggregate } from "../aggregates/session-aggregate";
import type { SessionRepository } from "../abstractions/session-repository";
import type { SessionService } from "../abstractions/session-service";
import type { CreateSessionDto } from "../dtos/create-session-dto";
import type { CreateSessionMessageDto } from "../dtos/create-session-message-dto";
import type { UpdateSessionDto } from "../dtos/update-session-dto";
import type { InferenceEngine } from "../../../../shared/abstractions/inference-engine";
import { Session } from "../entities/session";
import { SessionMessage } from "../entities/session-message";
import { SessionAggregate as SessionAggregateClass } from "../aggregates/session-aggregate";
import { NotFoundException } from "../../../../shared/exceptions/not-found-exception";

export class SessionServiceImpl implements SessionService {
  constructor(
    private readonly sessionRepository: SessionRepository,
    private readonly inferenceEngine: InferenceEngine,
  ) {}

  async create(sessionDto: CreateSessionDto): Promise<Session> {
    const session = new Session({
      name: sessionDto.name,
      presetId: sessionDto.presetId,
    });
    await this.inferenceEngine.createConversation({
      conversationId: session.id,
      systemPrompt: session.presetId,
    });
    try {
      return await this.sessionRepository.create(session);
    } catch (error) {
      await this.inferenceEngine.dropConversation(session.id).catch(() => undefined);
      throw error;
    }
  }

  async get(): Promise<Session[]> {
    return this.sessionRepository.get();
  }

  async getOneById(sessionId: string): Promise<SessionAggregate> {
    const session = await this.sessionRepository.getOneById(sessionId);
    if (session === null) {
      throw new NotFoundException(`Session not found: ${sessionId}`);
    }
    const messages = await this.sessionRepository.getSessionMessages(sessionId);
    return new SessionAggregateClass(session, messages);
  }

  async updateById(
    sessionId: string,
    updateDto: UpdateSessionDto,
  ): Promise<Session> {
    const aggregate = await this.getOneById(sessionId);
    if (updateDto.name !== undefined) {
      aggregate.name = updateDto.name;
      aggregate.touch();
    }
    const updated = await this.sessionRepository.updateById(
      sessionId,
      aggregate,
    );
    if (updated === null) {
      throw new NotFoundException(`Session not found: ${sessionId}`);
    }
    return updated;
  }

  async endById(sessionId: string): Promise<Session> {
    const aggregate = await this.getOneById(sessionId);
    await this.inferenceEngine.dropConversation(aggregate.id);
    aggregate.end();
    const updated = await this.sessionRepository.updateById(
      sessionId,
      aggregate,
    );
    if (updated === null) {
      throw new NotFoundException(`Session not found: ${sessionId}`);
    }
    return updated;
  }

  async deleteById(sessionId: string): Promise<void> {
    await this.inferenceEngine.dropConversation(sessionId).catch(() => undefined);
    await this.sessionRepository.deleteById(sessionId);
  }

  async createSessionMessage(
    sessionId: string,
    dto: CreateSessionMessageDto,
  ): Promise<SessionMessage> {
    const existingSession = await this.getOneById(sessionId);
    const message = new SessionMessage({
      sessionId,
      author: dto.author,
      content: dto.content,
      timestamp: dto.timestamp,
      errorMessage: dto.errorMessage,
      interrupted: dto.interrupted,
    });
    const { imagePath, audioPath } = await this.sessionRepository.persistMedia(
      sessionId,
      message.id,
      {
        imageBase64: dto.imageBase64,
        audioBase64: dto.audioBase64,
      },
    );

    const persistedMessage = new SessionMessage({
      id: message.id,
      sessionId: message.sessionId,
      author: message.author,
      content: message.content,
      timestamp: message.timestamp,
      imagePath,
      audioPath,
      errorMessage: message.errorMessage,
      interrupted: message.interrupted,
    });

    await this.sessionRepository.createSessionMessage(
      sessionId,
      persistedMessage,
    );
    existingSession.messageCount += 1;
    existingSession.touch();
    await this.sessionRepository.updateById(sessionId, existingSession);
    return persistedMessage;
  }

  async replaceMessage(
    sessionId: string,
    messageId: string,
    message: SessionMessage,
  ): Promise<SessionMessage> {
    const existingSession = await this.getOneById(sessionId);
    await this.sessionRepository.replaceSessionMessage(
      sessionId,
      messageId,
      message,
    );
    existingSession.touch();
    await this.sessionRepository.updateById(sessionId, existingSession);
    return message;
  }
}
